sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/InMemorySorter.java

```diff
@@ -0,0 +1,169 @@
+ * will reject additional pairs.
+ */
+public class InMemorySorter implements Sorter {
+  /** {@code Options} contains configuration of the sorter. */
+  public static class Options implements Serializable {
+    private int memoryMB = 100;
+
+    /** Sets the size of the memory buffer in megabytes. */
+    public void setMemoryMB(int memoryMB) {
+      this.memoryMB = memoryMB;
+    }
+
+    /** Returns the configured size of the memory buffer. */
+    public int getMemoryMB() {
+      return memoryMB;
+    }
+  }
+
+  /** The comparator to use to sort the records by key. */
+  private static final Comparator<byte[]> COMPARATOR = UnsignedBytes.lexicographicalComparator();
+
+  /** How many bytes per word in the running JVM. Assumes 64 bit/8 bytes if unknown. */
+  private static final int NUM_BYTES_PER_WORD = getNumBytesPerWord();
+
+  /**
+   * Estimate of memory overhead per KV record in bytes not including memory associated with keys
+   * and values.
+   *
+   * <ul>
+   * <li> Object reference within {@link ArrayList} (1 word),
+   * <li> A {@link KV} (2 words),
+   * <li> Two byte arrays (2 words for array lengths),
+   * <li> Per-object overhead (JVM-specific, guessing 2 words * 3 objects)
+   * </ul>
+   */
+  private static final int RECORD_MEMORY_OVERHEAD_ESTIMATE = 11 * NUM_BYTES_PER_WORD;
+
+  /** Maximum size of the buffer in bytes. */
+  private int maxBufferSize;
+
+  /** Current number of stored bytes. Including estimated overhead bytes. */
+  private int numBytes;
+
+  /** Whether sort has been called. */
+  private boolean sortCalled;
+
+  /** The stored records to be sorted. */
+  private ArrayList<KV<byte[], byte[]>> records = new ArrayList<KV<byte[], byte[]>>();
+
+  /** Private constructor. */
+  private InMemorySorter(Options options) {
+    maxBufferSize = options.getMemoryMB() * 1024 * 1024;
+  }
+
+  /** Create a new sorter from provided options. */
+  public static InMemorySorter create(Options options) {
+    return new InMemorySorter(options);
+  }
+
+  @Override
+  public void add(KV<byte[], byte[]> record) {
+    boolean successful = addIfRoom(record);
+    if (!successful) {
+      throw new IllegalStateException("No space remaining for in memory sorting");
+    }
+  }
+
+  /** Adds the record is there is room and returns true. Otherwise returns false. */
+  public boolean addIfRoom(KV<byte[], byte[]> record) {
+    if (sortCalled) {
+      throw new IllegalStateException("Records can only be added before sort().");
+    }
+
+    int recordBytes = estimateRecordBytes(record);
+    if (roomInBuffer(numBytes + recordBytes, records.size() + 1)) {
+      records.add(record);
+      numBytes += recordBytes;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public Iterable<KV<byte[], byte[]>> sort() {
+    if (sortCalled) {
+      throw new IllegalStateException("sort() can only be called once.");
```

Comment:
It does not matter with current use, but I wonder if it's okay to call `sort()` twice. It's just a no-op if already called?